### ISSUE TITLE
E-filing: Allow For More Than Three Iterations Of Document Upload

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8020]
As a Petitioner/Practitioner, I need to be able to accurately indicate which iteration of a document has been filed so that it can be correctly interpreted by the Court.

Currently, the UI for uploading a document to DAWSON as a petitioner or practitioner allows for only three iterations of a given document. In many cases, there are more than three iterations filed, but no method for indicating exactly which iteration past three a given upload represents. Currently these later iterations must be edited in the QC process by a docket clerk.

We would like a way for e-filing parties to indicate the correct iteration number of a filed document when filing said document.

## Acceptance Criteria
- Petitioner/Pracitioner efiling UI provides a way to indicate iterations of a given document beyond the third.
- Document hyperlink on the Docket Record in DAWSON accurately reflects the ordinal iteration of a given document.
- Document cover sheet when viewing the document in DAWSON accurately reflects the iteration of a given document.
- This applies to all documents for which we currently support multiple iterations (Amended, Amendment, Stipulation of Facts, Supplement, Supplemental)

## IRS API Considerations
Do these changes impact the IRS API?

## Security Considerations
 - [ ] Does this work make you nervous about privacy or security?
 - [ ] Does this work make major changes to the system?
 - [ ] Does this work implement new authentication or security controls?
 - [ ] Does this work create new methods of authentication, modify existing security controls, or explicitly implement any security or privacy features?

## Notes

![image.png](https://images.zenhubusercontent.com/60e7a3b441280d0c68e72525/e21e50f4-6249-4497-a97b-3f15acac64aa)

* Is free text entry by parties an option?
* Can this be automated?

## Tasks
- [x] Change Radio Buttons of First, Second, Third to a dropdown of 1-15 + "Other" https://github.com/flexion/ef-cms/pull/9873
- [x] Conditionally display the iteration input if "Other" is selected https://github.com/flexion/ef-cms/pull/9873
- [x] Question: What should the validation error be when the other iteration value entered is more than 3 characters? A: At standup 3/7, UX was good with "Maximum iteration value is 999."
- [x] For document types that have a populated ordinal value (NonStandard F, G, and I), pass the documentTitle through the ordinal to english translator (or ordinal suffix appender) - a helper would probably work best. https://github.com/flexion/ef-cms/pull/9873
- [x] If props contains otherIteration, use that instead of ordinalValue in generating the document title https://github.com/flexion/ef-cms/pull/9873
- [x] Styling of the iteration input https://github.com/flexion/ef-cms/pull/9873
- [x] Validation of the iteration input to be maximum of 3 length (999) in joi https://github.com/flexion/ef-cms/pull/9873
- [x] Validation that the iteration input be required if "Other" is selected https://github.com/flexion/ef-cms/pull/9873
- [x] Fix issue where "Other" with an otherIteration is provided, then the user goes back and changes iteration to another drop down value like "6". DocumentTitle should update. https://github.com/flexion/ef-cms/pull/9873
- [x] If "Other" with otherIteration is provided, then the user goes back and changes iteration to another drop down value like "6" but then goes back to "Other", the otherIteration field should be cleared. https://github.com/flexion/ef-cms/pull/9873
- [x] Question: casing of something like "Three Hundred And Forty-fifth" okay? Or should word after hyphen also be capitalized?

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios to meet all acceptance criteria have been written
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] Deployed to any Flexion environment for UX Review
 - [ ] Deployed to the Court's migration environment

**Comment by swongCO** [comment]
**UX Notes**

![My Cases - File a doc - iterations 1.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/f0ddb49d-fc9a-4aaa-b3a1-6ede1d902ada)

![My Cases - File a doc - iterations 2.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/b51addad-87ab-44b0-a425-7476a8bf5eee)

- [ ] Replace the three radio btns with a dropdown selector
- [ ] The dropdown list: 1-15, Other (at the bottom)
- [ ] If Other is selected, display an additional text field that accepts up to 4 numbers
- [ ] Display standard validation errors for the text field

**Comment by ttlenard** [comment]
## Test Cases

#### 1) Petitioner files a document in which we support multiple iterations; Petitioner can select an iteration beyond the 3rd iteration.

- Log in as a Petitioner that has an active case, and the petition has already been served to the IRS.
- Click on the Docket number of the case in your My cases list
- Click File a Document button
- Select one of the documents that supports multiple iterations (Amended, Amendment, Stipulation of Facts, Supplement, Supplemental)

Expected Results:
- [x] **NEW FUNCTIONALITY** - There is a new dropdown available that includes a numbered list (1-15) and the value "other".
- [x] **NEW FUNCTIONALITY** - If other is selected, a separate dropdown appears and the user can type in a number value (text is not supported, only number values).
- [x] **NEW FUNCTIONALITY** - Any number (up to 3 digits) can be entered in the number box .



#### 2) Document title on the File Your Document(s) screen accurately reflects the ordinal iteration selected from the new iteration dropdown menu.

- Continuing from the previous test
- Select a number from the dropdown list (do not choose other yet)
- Click Continue

Expected Results:
- [x] **NEW FUNCTIONALITY** - The ordinal iteration you selected from the dropdown menu is accurately titled on the File Your Document Screen. (Ex: if selected 14, it will say Fourteenth in front of the document title)


#### 3) Document title on the Review Your Filing screen accurately reflects the ordinal iteration selected from the new iteration dropdown menu.

- Continuing from the previous test
- Click on the Review Filing button

Expected Results:
- [x] **NEW FUNCTIONALITY** - The ordinal iteration you selected from the dropdown menu is accurately titled on the Review Your Filing Screen. (Ex: if selected 14, it will say Fourteenth in front of the document title)


#### 4) Document filed on the docket record accurately reflects the ordinal iteration selected from the new iteration dropdown menu.

- Continuing from the Previous Test
- Click Submit your filing

Expected Results:
- [x] **NEW FUNCTIONALITY** - The ordinal iteration you selected from the dropdown menu is accurately titled in the Filings and Proceedings Column. (Ex: if selected 14, it will say Fourteenth in front of the document title)


#### 5) Document Coversheet accurately reflects the ordinal iteration selected from the new iteration dropdown menu.

- Continuing from the previous test
- Click on the Document URL to open the document for viewing.

Expected Results
- [x] **NEW FUNCTIONALITY** - The ordinal iteration you selected from the dropdown menu is accurately titled on the document coversheet.


#### 6) Petitioner Selects Other from the iteration dropdown menu; Types in a number above 15; Ordinal iteration is reflected appropriately on File a Document, Review, and Docket Record Screens.

- Continuing from the previous tests
- Click on the File a Document button in your case
- Select one of the documents that supports multiple iterations (Amended, Amendment, Stipulation of Facts, Supplement, Supplemental)
- Select Other
- Enter in a number value over 15, but not to exceed 3 digits
- Click Continue (there will be 3 things to check after clicking continue...)

Expected Results:
- [x] **NEW FUNCTIONALITY** - On the File Your Document Screen, the ordinal iteration you entered in the Number entry box is accurately reflected in the Document title on this screen
- [x] **NEW FUNCTIONALITY** - On the Review Your Filing Screen, the ordinal iteration you entered in the Number entry box is accurately reflected in the Document title on this screen 
- [x] **NEW FUNCTIONALITY** - On the Docket Record Screen, the ordinal iteration you entered in the Number entry box is accurately reflected in the Document title on the docket record.
- [x] **NEW FUNCTIONALITY** - On the Document Coversheet, the ordinal iteration you entered in the Number entry box is accurately reflected in the Document title.


#### 7) Repeat test case 1-6 by testing with all of the following document types (Amended, Amendment, Stipulation of Facts, Supplement, Supplemental)

Expected Results:
- [x] All document types listed reflect the appropriate ordinal iteration selected or entered.


#### 8) Repeat the Test Cases above as a Private Practitioner

- Log in as a Private Practitioner
- Click on one of your cases from the My Cases Screen
- Click on File a document
- Select one of the documents that supports multiple iterations (Amended, Amendment, Stipulation of Facts, Supplement, Supplemental)

Expected Results:
- [x] Test case 1 passes with Private Practitioner
- [x] Test case 2 passes with Private Practitioner
- [x] Test case 3 passes with Private Practitioner
- [x] Test case 4 passes with Private Practitioner
- [x] Test case 5 passes with Private Practitioner
- [x] Test case 6 passes with Private Practitioner
- [x] Test case 7 passes with Private Practitioner


#### 9) Repeat the Test Cases above as an IRS Practitioner

- Log in as a IRS Practitioner
- Click on one of your cases from the My Cases Screen
- Click on File a document
- Select one of the documents that supports multiple iterations (Amended, Amendment, Stipulation of Facts, Supplement, Supplemental)

Expected Results:
- [x] Test case 1 passes with an IRS Practitioner
- [x] Test case 2 passes with an IRS Practitioner
- [x] Test case 3 passes with an IRS Practitioner
- [x] Test case 4 passes with an IRS Practitioner
- [x] Test case 5 passes with an IRS Practitioner
- [x] Test case 6 passes with an IRS Practitioner
- [x] Test case 7 passes with an IRS Practitioner

**Comment by cholly75** [comment]
@rosiehsmith - we can probably go ahead and cap the max iteration value at 999

**Comment by swongCO** [comment]
**UX Review feedback**

- disallow "." from being entered in the Iteration/Other field
- update empty state message for Iteration field to "Enter an iteration number."

**Comment by ttlenard** [comment]
Testing Feedback @rosiehsmith 

1. I noticed that if I selected Other from the dropdown menu, and then attempted to type in a number with my keyboard's keypad, it wouldn't register in the field. I could only use my number keys on the top row of my keyboard.

2. Also, once I entered in a number value, I could not backspace it or delete it. The only way I could get it to delete was to highlight the number, right click and select "cut".
